pkg/certificate/cert_manager/provider.go

```diff
@@ -0,0 +1,254 @@
+		pub := cert.PublicKey.(*rsa.PublicKey)
+		if !key.PublicKey.Equal(pub) {
+			return errors.New("private key(RSA) does not match the public key in the certificate")
+		}
+	case *ecdsa.PrivateKey:
+		pub := cert.PublicKey.(*ecdsa.PublicKey)
+		if !key.PublicKey.Equal(pub) {
+			return errors.New("private key(ECDSA) does not match the public key in the certificate")
+		}
+	default:
+		return fmt.Errorf("unsupported private key type: %T", key)
+	}
+
+	return nil
+}
+
+func (cm *Provider) ValidateCertificateSecret(
+	ctx context.Context,
+	secretName string,
+	namespace string,
+	_ *interfaces.Config) (bool, error) {
+	secret := &corev1.Secret{}
+	err := cm.Get(ctx, client.ObjectKey{Name: secretName, Namespace: namespace}, secret)
+	if err != nil {
+		return false, nil
+	}
+
+	certificateData, exists := secret.Data["tls.crt"]
+	if !exists {
+		return false, errors.New("certificate not found in secret")
+	}
+
+	decodeCertificatePem, _ := pem.Decode(certificateData)
+	if decodeCertificatePem == nil {
+		return false, errors.New("failed to decode PEM block")
+	}
+
+	privateKeyData, keyExists := secret.Data["tls.key"]
+	if !keyExists {
+		return false, errors.New("private key not found in secret")
+	}
+
+	parseCert, err := x509.ParseCertificate(decodeCertificatePem.Bytes)
+	if err != nil {
+		return false, fmt.Errorf("failed to parse certificate: %w", err)
+	}
+
+	if parseCert.NotAfter.Before(time.Now()) {
+		return false, errors.New("certificate has expired")
+	}
+
+	privateKey, err := parsePrivateKey(privateKeyData)
+	if err != nil {
+		return false, fmt.Errorf("failed to parse private key: %w", err)
+	}
+
+	if checkKeyPairErr := checkKeyPair(parseCert, privateKey); checkKeyPairErr != nil {
+		return false, fmt.Errorf("private key does not match certificate: %w", checkKeyPairErr)
+	}
+
+	return true, nil
+}
+
+func (cm *Provider) EnsureCertificateSecret(
+	ctx context.Context,
+	secretName string,
+	namespace string,
+	cfg *interfaces.Config) error {
+	issuerName, found := cfg.ExtraConfig["issuerName"].(string)
+	if !found || len(strings.TrimSpace(issuerName)) == 0 {
+		return errors.New("issuerName is missing in ExtraConfig")
+	}
+	issuerKind, found := cfg.ExtraConfig["issuerKind"].(string)
+	if !found || len(strings.TrimSpace(issuerKind)) == 0 {
+		return errors.New("issuerKind is missing in ExtraConfig")
+	}
+
+	checkCertSecret, valErr := cm.ValidateCertificateSecret(ctx, secretName, namespace, cfg)
+	if valErr != nil {
+		return fmt.Errorf("invalid certificate secret: %s present in namespace: %s, please delete and try again.\nError: %s",
+			secretName, namespace, valErr)
+	}
+	if checkCertSecret {
+		return fmt.Errorf("valid certificate secret: %s already present in namespace: %s , skipping Certificate creation",
+			secretName, namespace)
+	}
+	log.Printf("certificate secret: %s not present in namespace: %s , creating new Certificate",
+		secretName, namespace)
+
+	return cm.createCertificate(ctx, secretName, namespace, cfg)
```

Comment:
If the certManager is still in progress of creating the secret, so the ValidateCertificateSecret may fail to get/read the certificate, so etcd-operator will create the certificateResource again? It looks like a race condition.

Reply:
There should be official Cert-manager API for users (etcd-operator in this case) to check the progress.

Reply:
https://cert-manager.io/docs/usage/certificate/#inner-workings-diagram-for-developers

We should read the Certificate resource, and check the field `reason`,
- `Pending`
- `Issued`